File: script/UTSA_AP_MORTGAGE.py
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from scipy.stats import linregress

class APRateSpreadAnalysis:
    def __init__(self, num_bins=20, x_bounds=[0, 1000]):
        self.num_bins = num_bins
        self.x_bounds = x_bounds

    def load_data(self):
        self.ap_data = pd.read_csv('data/UTSA_AP_BY_SCHOOL.csv')
        self.rate_spread_data = pd.read_csv('data/RATESPREADLEA.csv')
        self.ap_data['LEAID'] = self.ap_data['LEAID'].astype(str).str.lower()
        self.rate_spread_data['leaid'] = self.rate_spread_data['leaid'].astype(str).str.lower()
        self.rate_spread_data.rename(columns={'leaid': 'LEAID', 'avg_rate_spread': 'RateSpread'}, inplace=True)

    def plot_rate_spread_bins(self):
        # Make sure RateSpread is used after renaming
        if 'RateSpread' not in self.rate_spread_data.columns:
            self.rate_spread_data.rename(columns={'avg_rate_spread': 'RateSpread'}, inplace=True)

        # Ensure unique entries for rate spread per LEAID
        self.rate_spread_data = self.rate_spread_data.groupby('LEAID').agg({'RateSpread': 'mean'}).reset_index()
        self.ap_data = self.ap_data.groupby('LEAID').agg({'AP': 'sum'}).reset_index()
        
        # Create bins for rate spread with the right edge included
        self.rate_spread_data['RateSpreadBin'] = pd.cut(self.rate_spread_data['RateSpread'], bins=np.linspace(self.rate_spread_data['RateSpread'].min(), self.rate_spread_data['RateSpread'].max(), self.num_bins + 1), right=True, include_lowest=True)
        
        # Merge datasets on LEAID and sum AP by rate spread bins
        merged_data = pd.merge(self.rate_spread_data, self.ap_data, on='LEAID', how='inner')
        binned_ap_data = merged_data.groupby('RateSpreadBin')['AP'].sum().reset_index()
        
        # Plot the AP distribution across rate spread bins
        plt.figure(figsize=(12, 8))
        plt.bar(binned_ap_data['RateSpreadBin'].astype(str), binned_ap_data['AP'], color='skyblue')
        plt.xticks(rotation=45)
        plt.xlabel('Rate Spread Bins')
        plt.ylabel('Total AP Students')
        plt.title('Total AP Students by Rate Spread Bins')
        plt.grid(True)
        plt.tight_layout()  # Adjust layout to make room for label rotation
        plt.show()

# Use the class
if __name__ == "__main__":
    analysis = APRateSpreadAnalysis()
    analysis.load_data()
    analysis.plot_rate_spread_bins()

File: script/UTSA_AP_MORTGAGE.r
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
from scipy.stats import linregress

# Define the x-axis bounds
x_bounds = [0, 1000]

# Load the datasets
ap_data = pd.read_csv('../data/UTSA_AP_BY_SCHOOL.csv')
rate_spread_data = pd.read_csv('../data/RATESPREADLEA.csv')

# Ensure LEAID columns are strings and lower case for consistent merging
ap_data['LEAID'] = ap_data['LEAID'].astype(str).str.lower()
rate_spread_data['leaid'] = rate_spread_data['leaid'].astype(str).str.lower()
rate_spread_data.rename(columns={'leaid': 'LEAID'}, inplace=True)

# Sum the AP credits by school district (LEAID) and apply x-axis bounds
ap_summary = ap_data.groupby('LEAID')['AP'].sum().reset_index()
ap_summary = ap_summary[(ap_summary['AP'] >= x_bounds[0]) & (ap_summary['AP'] <= x_bounds[1])]

# Merge the datasets on LEAID
merged_data = pd.merge(ap_summary, rate_spread_data, on='LEAID', how='inner')

# Perform linear regression
regression_results = linregress(merged_data['AP'], merged_data['avg_rate_spread'])

# Unpack regression results
slope = regression_results.slope
intercept = regression_results.intercept
r_value = regression_results.rvalue
r_squared = r_value**2

# Create regression line for plotting
regression_line = slope * merged_data['AP'] + intercept

# Update plot with regression information and x_bounds
plt.figure(figsize=(10, 6))
plt.errorbar(merged_data['AP'], merged_data['avg_rate_spread'], yerr=merged_data['standarddeviation'], fmt='o', ecolor='red', capsize=5)
plt.plot(merged_data['AP'], regression_line, color='blue', label=f'Linear fit: y = {slope:.2f}x + {intercept:.2f}\n$R^2 = {r_squared:.4f}$')
plt.xlabel('Total AP per School District')
plt.ylabel('Average Rate Spread')
plt.title('AP Scores vs. Mortgage Rate Spread by School District')
plt.xlim(x_bounds)
plt.legend()
plt.grid(True)
plt.show()

# Output regression parameters and R^2
print(f"Linear Regression Equation: y = {slope:.2f}x + {intercept:.2f}")
print(f"Slope: {slope:.2f}")
print(f"Intercept: {intercept:.2f}")
print(f"R-squared: {r_squared:.4f}")
